File: java/google/registry/ui/js/registrar/host.js
// Copyright 2017 The Nomulus Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

goog.provide('registry.registrar.Host');

goog.require('registry.registrar.XmlResourceComponent');
goog.require('registry.soy.registrar.host');
goog.require('registry.soy.registrar.hostepp');

goog.forwardDeclare('registry.registrar.Console');



/**
 * CRUD for EPP host objects.
 * @param {!registry.registrar.Console} console
 * @constructor
 * @extends {registry.registrar.XmlResourceComponent}
 * @final
 */
registry.registrar.Host = function(console) {
  registry.registrar.Host.base(
      this, 'constructor',
      registry.soy.registrar.host.item,
      registry.soy.registrar.hostepp,
      console);
};
goog.inherits(registry.registrar.Host,
              registry.registrar.XmlResourceComponent);


/** @override */
registry.registrar.Host.prototype.processItem = function() {
  this.model.item = this.model['epp']['response']['resData']['host:infData'];
  if (this.model.item['host:addr']) {
    if (!goog.isArray(this.model.item['host:addr'])) {
      this.model.item['host:addr'] = [this.model.item['host:addr']];
    }
  } else {
    this.model.item['host:addr'] = [];
  }
};


/** @override */
registry.registrar.Host.prototype.setupEditor = function(objArgs) {
  this.typeCounts['host-addr'] =
      objArgs.item['host:addr'] ? objArgs.item['host:addr'].length : 0;
  this.addRemBtnHandlers('host-addr', goog.bind(function() {
    return 'host:addr[' + this.typeCounts['host-addr'] + ']';
  }, this));

  this.formInputRowRemovable(document.querySelectorAll('input[readonly]'));
};


/** @override */
registry.registrar.Host.prototype.prepareCreate = function(params) {
  params.nextId = params.item['host:name'];
  return registry.soy.registrar.hostepp.create(params).toString();
};


/** @override */
registry.registrar.Host.prototype.prepareUpdate = function(params) {
  var form = params.item;
  var addAddrs = [];
  var remAddrs = [];
  if (form['host:addr']) {
    var oldAddrs = form['host:oldAddr'] || [];
    var newAddrs = form['host:addr'];
    var length = Math.max(oldAddrs.length, newAddrs.length);
    for (var i = 0; i < length; i++) {
      if (i >= oldAddrs.length) {
        addAddrs.push(newAddrs[i]['value']);
      } else if (i >= newAddrs.length) {
        remAddrs.push(oldAddrs[i]['value']);
      } else {
        if (newAddrs[i]['value'] == oldAddrs[i]['value']) {
          // Do nothing.
        } else if (newAddrs[i]['value'] == '') {
          remAddrs.push(oldAddrs[i]['value']);
        } else {
          remAddrs.push(oldAddrs[i]['value']);
          addAddrs.push(newAddrs[i]['value']);
        }
      }
    }
  }
  params.addAddrs = addAddrs;
  params.remAddrs = remAddrs;
  params.nextId = form['host:chgName'];
  return registry.soy.registrar.hostepp.update(params).toString();
};

File: core/src/main/java/google/registry/model/host/Host.java
// Copyright 2017 The Nomulus Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package google.registry.model.host;

import google.registry.model.EppResource.ForeignKeyedEppResource;
import google.registry.model.annotations.ExternalMessagingName;
import google.registry.persistence.VKey;
import google.registry.persistence.WithVKey;
import javax.persistence.Access;
import javax.persistence.AccessType;
import javax.persistence.Entity;
import javax.persistence.Id;
import javax.persistence.Index;
import javax.persistence.Table;

/**
 * A persistable Host resource including mutable and non-mutable fields.
 *
 * <p>The {@link Id} of the Host is the repoId.
 */
@Entity(name = "Host")
@Table(
    name = "Host",
    /*
     * A gin index defined on the inet_addresses field ({@link HostBase#inetAddresses} cannot be
     * declared here because JPA/Hibernate does not support index type specification. As a result,
     * the hibernate-generated schema (which is for reference only) does not have this index.
     *
     * <p>There are Hibernate-specific solutions for adding this index to Hibernate's domain model.
     * We could either declare the index in hibernate.cfg.xml or add it to the {@link
     * org.hibernate.cfg.Configuration} instance for {@link SessionFactory} instantiation (which
     * would prevent us from using JPA standard bootstrapping). For now, there is no obvious benefit
     * doing either.
     */
    indexes = {
      @Index(columnList = "hostName"),
      @Index(columnList = "creationTime"),
      @Index(columnList = "deletionTime"),
      @Index(columnList = "currentSponsorRegistrarId"),
      @Index(columnList = "dnsRefreshRequestTime")
    })
@ExternalMessagingName("host")
@WithVKey(String.class)
@Access(AccessType.FIELD) // otherwise it'll use the default if the repoId (property)
public class Host extends HostBase implements ForeignKeyedEppResource {

  @Override
  @Id
  @Access(AccessType.PROPERTY) // to tell it to use the non-default property-as-ID
  public String getRepoId() {
    return super.getRepoId();
  }

  @Override
  public VKey<Host> createVKey() {
    return VKey.create(Host.class, getRepoId());
  }

  @Override
  public Builder asBuilder() {
    return new Builder(clone(this));
  }

  /** A builder for constructing {@link Host}, since it is immutable. */
  public static class Builder extends HostBase.Builder<Host, Builder> {
    public Builder() {}

    private Builder(Host instance) {
      super(instance);
    }

    public Builder copyFrom(HostBase hostBase) {
      return setCreationRegistrarId(hostBase.getCreationRegistrarId())
          .setCreationTime(hostBase.getCreationTime())
          .setDeletionTime(hostBase.getDeletionTime())
          .setHostName(hostBase.getHostName())
          .setInetAddresses(hostBase.getInetAddresses())
          .setLastTransferTime(hostBase.getLastTransferTime())
          .setLastSuperordinateChange(hostBase.getLastSuperordinateChange())
          .setLastEppUpdateRegistrarId(hostBase.getLastEppUpdateRegistrarId())
          .setLastEppUpdateTime(hostBase.getLastEppUpdateTime())
          .setPersistedCurrentSponsorRegistrarId(hostBase.getPersistedCurrentSponsorRegistrarId())
          .setRepoId(hostBase.getRepoId())
          .setSuperordinateDomain(hostBase.getSuperordinateDomain())
          .setDnsRefreshRequestTime(hostBase.getDnsRefreshRequestTime())
          .setStatusValues(hostBase.getStatusValues());
    }
  }
}
